text/0000-prepare-operation.md

```diff
@@ -0,0 +1,374 @@
+3. [Supporting Cloud Native Buildpack utilities.](#Cloud-Native-Buildpacks-utilities)
+
+## Changes to `io.buildpacks` namespace
+
+### Namespace `io.buildpacks.defaults`
+
+This namespace declares default values that platforms should acknowledge. A majority of properties should be spec'd but we would allow for unspec'd additional properties.
+
+The criteria for spec'd properties in this namespace would be based on whether they affect inputs to the lifecycle phases. For example, the order and environment variables are inputs to the lifecycle. The `builder`, in contrast, is not an input to the lifecycle so it wouldn't meet the criteria.
+
+Therefore the following changes have been made from the [latest schema][pd-02]:
+
+ - Remove `builder`
+
+[pd-02]: https://github.com/buildpacks/spec/blob/extensions/project-descriptor/0.2/extensions/project-descriptor.md#iobuildpacks-optional
+
+### Version `io.buildpacks` namespace
+
+Given that the project descriptor's `io.buildpacks` namespace is directly tied to the platform by the fact that it's the platform that needs to consume and parse it, it makes sense to unify the version of the schema with that of the platform API.
+
+
+### Schema
+
+See [Platform Spec changes](#Platform-Spec).
+
+#### Usages Examples
+
+###### Multi-Platform (pack + kpack)
+
+In the following example we show how various platform namespaces can have their own configuration but properties can be promoted to `io.buildpacks.defaults` if the user's intent is to apply it to all platforms.
+
+```toml
+[io.buildpacks]
+schema-version = "0.2"
+
+###
+# common buildpacks config
+##
+
+[io.buildpacks.defaults]
+exclude = ["some-files/**"]
+
+##
+# pack config
+##
+
+[io.buildpacks.pack]
+schema-version = "0.1"
+builder = "cnbs/sample-builder:bionic"
+image = "my-app"                        # (example only)
+run-image = "my-run-image:latest"       # (example only)
+
+##
+# kpack config
+##
+
+[com.vmware.kpack]
+schema-version = "0.1"
+
+[[com.vmware.kpack.build.env]]
+name="CUSTOM_ENV"
+value="SOME_VALUE"
+```
+
+Notice:
+
+1. Both `pack` and `kpack` are expected to apply `io.buildpacks.defaults.exclude`.
+2. `io.buildpacks.pack` has additional properties that are specific to `pack` only.
+
+## Prepare phase
+
+A `prepare` phase would be a new lifecycle phase that gets executed before `create` or `detect` phase. Similar to the `rebaser`, it is outside of the standard [Build operation][build-operation].
+
+[build-operation]: https://github.com/buildpacks/spec/blob/main/platform.md#build
+
+#### Responsibilities
+
+At minimum, the expected reponsibility of the `preparer` would be to:
+
+ - Apply the requested configuration.
+
+#### Inputs
+
+The `prepare` phase would take all the same inputs as `creator` plus the addition of a path to `project.toml`.
+
+#### Outputs
+
+The `prepare` phase may affect the file system, and mutation or create files that would be consumed be futher phases.
```

Comment:
I think we should elaborate here a bit. What are the limits? What volumes should be mounted? Is this phase intended/allowed to happen in a different container?

Thinking out loud a bit on use cases I can see leveraging a prepare phase

- Download source code (from a signed url, for instance, provided by the platform)
- Downloading previous build cache (again, from a signed url or perhaps copying from a k8s style PVC)
- Generating a `project.toml` for projects that don't have one (for platform consistency, maybe)

To support those, we would need to mount `/workspace` and `/cache`.

Reply:
> What volumes should be mounted? 

Volumes aren't really part of the spec. It sounds like you may be asking from a specific PoV. We could elaborate through examples but volumes aren't referenced because they aren't really required from a platform -> lifecycle interface.

> Is this phase intended/allowed to happen in a different container?

[This answer here](https://github.com/buildpacks/rfcs/pull/202/files/99fc145c93f07e045e62b9c5d42405ecffc53afc#r801953808) may shed some light on the question. It is intended to run on a separate container although not technically required to do so. Most platforms would execute their own prepare binary/container but some may rely on the [builder-provided `preparer`](https://github.com/buildpacks/rfcs/blob/99fc145c93f07e045e62b9c5d42405ecffc53afc/text/0000-prepare-operation.md?plain=1#L147-L160).